tensilelite/rocisa/rocisa/include/helper.hpp

```diff
@@ -47,50 +49,87 @@ inline std::string getGfxNameTuple(const IsaVersion& isaVersion)
            + int_to_hex[isaVersion[2]];
 }
 
-inline std::pair<int, std::string>
-    run(const std::vector<char*>& cmd, const std::string& input, bool debug = false)
+#define CHECK_COMGR(call)                                                                  \
+    do                                                                                     \
+    {                                                                                      \
+        amd_comgr_status_t status = call;                                                  \
+        if(status != AMD_COMGR_STATUS_SUCCESS)                                             \
+        {                                                                                  \
+            const char* status_string = nullptr;                                           \
+            amd_comgr_status_string(status, &status_string);                               \
+            throw std::runtime_error(std::string("COMGR call failed: ") + #call            \
+                                     + " (COMGR status: " + std::to_string(status) + ", "  \
+                                     + (status_string ? status_string : "Unknown") + ")"); \
+        }                                                                                  \
+    } while(0)
+
+inline std::pair<int, std::string> run(const std::string& input,
+                                       const std::string& gfxStr,
+                                       const char**       optionStr,
+                                       const size_t       optionSize,
+                                       bool               debug = false)
 {
-    int   p[2];
-    pid_t pid;
-    if(pipe(p) == -1)
-    {
-        throw std::runtime_error("cmd failed!");
-    }
+    amd_comgr_data_t        data;
+    amd_comgr_data_set_t    dataSet;
+    amd_comgr_data_set_t    outputDataSet;
+    amd_comgr_action_info_t actionInfo;
+    std::string             log;
 
-    pid = fork();
-    if(pid == 0)
+    // Create data object for input assembly code
+    CHECK_COMGR(amd_comgr_create_data(AMD_COMGR_DATA_KIND_SOURCE, &data));
+    if(input.size() > 0)
     {
-        close(p[1]);
-        dup2(p[0], STDIN_FILENO);
-        if(!debug)
-        {
-            dup2(p[0], STDERR_FILENO);
-            dup2(p[0], STDOUT_FILENO);
-        }
-        close(p[0]);
-        execvp(cmd[0], cmd.data());
-        perror("execvp");
-        exit(1);
+        CHECK_COMGR(amd_comgr_set_data(data, input.size(), input.c_str()));
     }
-    else
+    CHECK_COMGR(amd_comgr_set_data_name(data, "input.s"));
+
+    // Create data set and add the input data
+    CHECK_COMGR(amd_comgr_create_data_set(&dataSet));
+    CHECK_COMGR(amd_comgr_create_data_set(&outputDataSet));
+    CHECK_COMGR(amd_comgr_data_set_add(dataSet, data));
```

Comment:
What happens if any of these get an error and throw? As I read it, every one of them leaks memory. 

You need to rewrite this so that your allocated entities are contained by a raii instance that will deallocate any allocated things on destruction. See #1951 for one (of many) ways you can do this to bridge too a c based API safely.

Reply:
I'll change to unique_ptr instead

Reply:
I'm not sure that is a good way: these have explicit apis for create/release and do not have the same semantics as delete with respect to nullptr afaict. But I'll have a look at what you come up with.